core/storage/btree.rs

```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
+        }
+        total_used -= page.num_frag_free_bytes() as usize;
 
-        let is_page_1 = {
-            let current_root = self.stack.top();
-            current_root.get().id == 1
-        };
+        total_used
+    }
+
+    fn estimate_cell_pointer_growth(&self) -> usize {
+        // In SQLite, we want to leave some space for future cell pointers
+        // Each cell pointer takes 2 bytes
+        const CELL_POINTER_SIZE: usize = 2;
+        const ESTIMATED_FUTURE_CELLS: usize = 4; // Leave room for a few more cells
+
+        CELL_POINTER_SIZE * ESTIMATED_FUTURE_CELLS
+    }
+
+    fn is_overflow(&self, page: &PageContent) -> bool {
+        if !page.overflow_cells.is_empty() {
+            return true;
+        }
+
+        let used_space = self.calculate_used_space(page);
+        let available_space = self.usable_space();
 
+        // Check if we're about to exceed usable space
+        // We also need some buffer space for future cell pointers
+        let required_space = used_space + self.estimate_cell_pointer_growth();
+        required_space > available_space
+    }
+
+    fn is_underflow(&self, page: &PageContent) -> bool {
+        // Root is special case - only underflow when empty with one child
+        let current_page = self.stack.top();
+        if current_page.get().id == self.root_page {
+            return page.cell_count() == 0 && !page.is_leaf();
+        }
+
+        let used_space = self.calculate_used_space(page);
+        let usable_space = self.usable_space();
+
+        // Underflow if used space is less than 50% of usable space
+        used_space < (usable_space / 2)
+    }
+
+    fn balance_root(&mut self) {
+        let current_root = self.stack.top();
+        let contents = current_root.get().contents.as_mut().unwrap();
+        let is_page_1 = current_root.get().id == 1;
         let offset = if is_page_1 { DATABASE_HEADER_SIZE } else { 0 };
-        let new_root_page = self.allocate_page(PageType::TableInterior, offset);
-        {
-            let current_root = self.stack.top();
-            let current_root_contents = current_root.get().contents.as_ref().unwrap();
 
-            let new_root_page_id = new_root_page.get().id;
-            let new_root_page_contents = new_root_page.get().contents.as_mut().unwrap();
-            if is_page_1 {
-                // Copy header
-                let current_root_buf = current_root_contents.as_ptr();
-                let new_root_buf = new_root_page_contents.as_ptr();
-                new_root_buf[0..DATABASE_HEADER_SIZE]
-                    .copy_from_slice(&current_root_buf[0..DATABASE_HEADER_SIZE]);
+        if self.is_underflow(contents) {
+            let child_page_id = contents.rightmost_pointer().unwrap();
+            let child_page = self
+                .pager
+                .read_page(child_page_id as usize)
+                .expect("This shouldn't have happened, child page not found");
```

Comment:
Better to propagate errors
```suggestion
            let child_page = self
                .pager
                .read_page(child_page_id as usize)?;
```

And I think we are not handling the case where child page is not loaded/locked?